### ISSUE TITLE
No type error when assigning from generic type parameter to remapped other type parameter

### ISSUE THREAD

**Issue by fatcerberus** [microsoft/TypeScript#31457]
**TypeScript Version:**  3.4.5 (wanted to try `@next` but couldn't figure out how to get VSCode to use it)

**Search Terms:**
* typeof no error
* (this was really hard to search for)

**Code**

```ts
type TypeOfResult = 'bigint' | 'boolean' | 'function' | 'number' | 'object' | 'string' | 'symbol' | 'undefined';
type TypeOf<K extends TypeOfResult> = {
	bigint: bigint,
	boolean: boolean,
	function: (...args: never[]) => unknown,
	number: number,
	object: object | null,
	string: string,
	symbol: symbol,
	undefined: undefined,
}[K];

function ofType<T, K extends TypeOfResult>(values: T[], ...types: K[]): TypeOf<K>[]
{
	return values.filter(it => true);  // no type error?!
}

const onlyTheStrings = ofType([ 1, "pig", 2, "cow", 3, "ape" ], 'string');
console.log(onlyTheStrings);  // bug!
```

**Expected behavior:**
The compiler should flag a type error for the `return` since `T` is not (provably) assignable to `TypeOf<K>`

**Actual behavior:**
No error; repro code compiles without issue despite the buggy filter and lack of type assertions.  TS thinks the return type at the call site is `string[]` but at runtime the non-strings are not filtered out.

**Playground Link:** 
https://typescript-play.js.org/#code/C4TwDgpgBAKuEHkBmAlCBnArgG2FAvFAOQBGAlgOZkB2wRUAPsSQPYvYQCG19TRSmagGNgZFj0bFqmALYkIAJ17EWJAFYQRyoumAKaFbehBz22wQBMISGhAtEA3AChQkWPGQAeANJQIAD2AIagt0d0hkNCxcAD4CKABvJwBIcipaAC4oNJpgABoU1nYuaiyijm4C5IFhUXEsgAoAOhbOBQp0LOoIADdFAG0AXQBKAjjBAGtqFgB3airpOUUu2XkFKtUNESzNzTwmaWxsKt19agos04MTk1ZsS9v2KstrWwssl5tuiwKAX37vINnE4aiIxNQoCwkHBIJ4YHkoL4AkEQmEYYhUBgcMAYg0epxsJgMFkYEMES0mq5iYihsMSR4kD4YkMnElkgoIMBMAoIfjCRgHFAoAB6YVQaZQKl+BQKFgKAD8AEInL8nE4hOJdJDqNgQDAABYQADKegMYUIUPRDX6UAAjAiAERgSgOhEAJkdGpmrqgAGZHZxIA6oIMETpTeciMNnBrqOhik1sCwKA1xLqDcaIx1o0LRdlMBRlUA

**Related Issues:**
Couldn't find any but I didn't really know what keywords to use either.

**Comment by ahejlsberg** [comment]
This is fixed by #30769 and your example is an error in 3.5.

**Comment by j-oliveras** [comment]
[Using newer TypeScript versions on VSCode](https://code.visualstudio.com/docs/typescript/typescript-compiling#_using-newer-typescript-versions).

**Comment by fatcerberus** [comment]
@ahejlsberg 

> This is fixed by #30769

Hm, that's interesting.  Reading through that PR it seems like the thing that must have fixed this case is:
> When an indexed access T[K] occurs on the source side of a type relationship, it resolves to a union type of the properties selected by T[K], but when it occurs on the target side of a type relationship, it now resolves to an intersection type of the properties selected by T[K]. Previously, the target side would resolve to a union type as well, which is unsound.

But that doesn't seem right - `T` and `TypeOf<K>` are not obviously related at all (while still in generic context, i.e. not yet resolved to concrete types) so it doesn't seem like union/intersection should make any difference?  Neither should be assignable to the other regardless of how the compiler relates them.

Not being argumentative, just trying to figure out what's going on is all. :smiley:

**Comment by fatcerberus** [comment]
...unless what it's doing is treating the return type as if it's `TypeOf<TypeOfResult>` (the constraint of `K`)?  Then in 3.4 that would resolve to `string | number | ...` (i.e. every possible primitive type) while in 3.5 it resolves to `string & number & ...` or in other words `never` so nothing is directly assignable to it.

That's probably what's happening.  I had assumed it treated `K` as a complete unknown quantity without considering it had a concrete constraint the compiler could look at instead.

**Comment by fatcerberus** [comment]
So I just have one more question.  This is the actual code I had written that led me to discover the issue:

```ts
ofType<K extends TypeOfResult>(...types: K[]): Query<TypeOf<K>>
{
	const typeSet = new Set<TypeOfResult>(types);
	return this.where(it => typeSet.has(typeof it));
}
```

Unlike the example given in the OP, *this* code is actually fine, as the filter is doing what it's supposed to do.  What I'm wondering: in TS 3.5, will there be a way to convince the compiler the above is okay, short of casting the return value to `any`?

**Comment by ahejlsberg** [comment]
> What I'm wondering: in TS 3.5, will there be a way to convince the compiler the above is okay, short of casting the return value to any?

I couldn't tell you without knowing the types of `where` and `Query`.

**Comment by fatcerberus** [comment]
https://github.com/fatcerberus/from.js/blob/master/src/from.ts#L331

`Query` is generic.  It's essentially a lazy-evaluated `Array#filter`.
